rs/state_manager/src/lib.rs

```diff
@@ -862,9 +862,6 @@ fn check_certifications_metadata_snapshots_and_states_metadata_are_consistent(
         .filter(|h| h.get() != 0)
         .collect::<Vec<_>>();
     debug_assert_eq!(certification_heights, snapshot_heights);
-    for h in states.states_metadata.keys() {
-        debug_assert!(states.certifications_metadata.contains_key(h));
-    }
```

Comment:
One thing I don't quite remember is whether we introduced these `debug_assert`s (the one here and the ones below) just to sanity check whether the implementation behaves according to our understanding, or whether we actually rely on these invariants somewhere. The former would mean adapting them according to the new behavior is fine, the latter might mean that this change will lead to violating an invariant some other part of the code relies on.

Reply:
This has been discussed in meetings. For the information of other viewers of the PR:
The these `debug_asserts` were introduced to ensure the consistency between in-memory states and states_metadata. If the asserts held,  it would be safe to make decisions on whether to insert in-memory states and states_metadata solely based on checking in-memory states.

Because this PR treats in-memory states and checkpoints separately when removing states. These `debug_asserts` do not hold anymore. As a result, this PR also adjusts several places so that we check in-memory states and states_metadata respectively when making decisions on whether to add one.